### PR TITLE
Remove Scratchpad from KernelManager

### PR DESCRIPTION
**Pull request by mzient**
A follow-up for #3670 

## Category:

**Refactoring** (*Redesign of existing code that doesn't affect functionality*)

## Description:
After dropping the use of preallocated scratchpad in KernelManager, it is quite pointless to keep ScratchpadAllocator instances in it - also, passing thread_idx to `Run` is no longer necessary, since it was only used for indexing scratchpads. As such, this PR removes those obsolete pieces.

### Tests
- [X] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [X] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2449

### CODE REVIEW

File: dali/kernels/kernel_manager.h [hunk 1]
```diff
@@ -189,124 +162,44 @@ class DLL_PUBLIC KernelManager {
...
-  inline ScratchSizes MaxScratchSizes() const {
-    ScratchSizes sizes;
-    for (size_t i = 0; i < sizes.size(); i++) {
-      sizes[i] = max_scratch_sizes[i];
+    if (!context.scratchpad) {
+      DynamicScratchpad scratchpad({}, AccessOrder(context.gpu.stream));
+      auto *old_scratchpad = context.scratchpad;
+      context.scratchpad = &scratchpad;
+      inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
+      context.scratchpad = old_scratchpad;
```
Comment (JanuszL) [inline dali/kernels/kernel_manager.h:195]:
```suggestion
      context.scratchpad = &scratchpad;
      inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
      context.scratchpad = nullptr;
```

Comment (mzient) [reply dali/kernels/kernel_manager.h:195]: :+1:

File: include/dali/core/call_at_exit.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  }
+  Callable callable;
+};
+
+}  // namespace detail
+
+template <typename Callable>
+detail::CallAtExit<Callable> AtBlockExit(Callable &&c) {
+  return detail::CallAtExit<Callable>(std::forward<Callable>(c));
+}
```
Comment (mzient) [inline include/dali/core/call_at_exit.h:40]: Copied from default_resources.cc

**Review by JanuszL — APPROVED** [review]

File: dali/operators/image/distortion/jpeg_compression_distortion_op_gpu.cu [hunk 3]
```diff
@@ -24,7 +24,7 @@ class JpegCompressionDistortionGPU : public JpegCompressionDistortion<GPUBackend
  public:
   explicit JpegCompressionDistortionGPU(const OpSpec &spec) : JpegCompressionDistortion(spec) {
     kmgr_.Initialize<JpegDistortionKernel>();
-    kmgr_.Resize<JpegDistortionKernel>(1, 1);
+    kmgr_.Resize<JpegDistortionKernel>(1);
```
Comment (jantonguirao) [inline dali/operators/image/distortion/jpeg_compression_distortion_op_gpu.cu:27]: remove initialize

File: dali/operators/signal/decibel/to_decibels_op_cpu.cc [hunk 4]
```diff
@@ -56,12 +56,11 @@ bool ToDecibels<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
   kernels::KernelContext ctx;
   auto in_shape = input.shape();
   int nsamples = input.num_samples();
-  auto nthreads = ws.GetThreadPool().NumThreads();
 
   TYPE_SWITCH(input.type(), type2id, T, (float), (
     using ToDbKernel = kernels::signal::ToDecibelsCpu<T>;
     kmgr_.Initialize<ToDbKernel>();
-    kmgr_.Resize<ToDbKernel>(nthreads, nsamples);
+    kmgr_.Resize<ToDbKernel>(nsamples);
```
Comment (jantonguirao) [inline dali/operators/signal/decibel/to_decibels_op_cpu.cc:63]: remove initialize

File: dali/operators/signal/fft/power_spectrum.cc [hunk 5]
```diff
@@ -63,15 +63,14 @@ bool PowerSpectrum<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
...
-  auto nthreads = ws.GetThreadPool().NumThreads();
 
   // Other types not supported for now
   using InputType = float;
   using OutputType = float;
   VALUE_SWITCH(in_shape.sample_dim(), Dims, FFT_SUPPORTED_NDIMS, (
     using FftKernel = kernels::signal::fft::Fft1DCpu<OutputType, InputType, Dims>;
     kmgr_.Initialize<FftKernel>();
-    kmgr_.Resize<FftKernel>(nthreads, nsamples);
+    kmgr_.Resize<FftKernel>(nsamples);
```
Comment (jantonguirao) [inline dali/operators/signal/fft/power_spectrum.cc:73]: remove initialize

File: dali/operators/signal/fft/spectrogram.cc [hunk 6]
```diff
@@ -204,8 +204,7 @@ bool SpectrogramImplCpu<time_major>::SetupImpl(std::vector<OutputDesc> &out_desc
     if (!w.raw_data()) w.set_pinned(false);
   }
 
-  kmgr_window_.Initialize<WindowKernel>();
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/signal/fft/spectrogram.cc:207]: Yes, it's done implicitly in Resize

File: include/dali/core/call_at_exit.h [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  ~CallAtExit() noexcept(false) {
+    callable();
+  }
+  Callable callable;
+};
+
+}  // namespace detail
+
+template <typename Callable>
+detail::CallAtExit<Callable> AtBlockExit(Callable &&c) {
```
Comment (jantonguirao) [inline include/dali/core/call_at_exit.h:38]:
```suggestion
detail::CallAtExit<Callable> AtScopeExit(Callable &&c) {
```
would be more understandable, I think

**Review by jantonguirao — APPROVED** [review]
LGTM, with some minor comments

File: dali/kernels/kernel_manager.h [hunk 8]
```diff
@@ -190,130 +163,45 @@ class DLL_PUBLIC KernelManager {
...
-      sizes[i] = max_scratch_sizes[i];
+    if (!context.scratchpad) {
+      DynamicScratchpad scratchpad({}, AccessOrder(context.gpu.stream));
+      context.scratchpad = &scratchpad;
+      auto finally = AtBlockExit([&]() {
+        context.scratchpad = nullptr;
+      });
+      inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
+    } else {
+      inst.get<Kernel>().Run(context, std::forward<OutInArgs>(out_in_args)...);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/kernel_manager.h:199]: This `finally` would go out of scope too early.

**Pull request merged**